src/main/java/hudson/tasks/junit/ClassResult.java

```diff
@@ -146,7 +144,7 @@ public Object getDynamic(String name, StaplerRequest req, StaplerResponse rsp) {
 
     @Exported(name="child")
     @Override
-    public List<CaseResult> getChildren() {
+    public Collection<CaseResult> getChildren() {
```

Comment:
this is a breaking change as other plugins have this as a list where its now actually a set, see e.g. 
https://github.com/jenkinsci/junit-attachments-plugin/blob/225d1e03b0a8cae4661ca425e71b69757723318c/src/test/java/hudson/plugins/junitattachments/AttachmentPublisherPipelineTest.java#L81

Build result showing test failures:
https://ci.jenkins.io/blue/organizations/jenkins/Tools%2Fbom/detail/PR-3394/1/tests

Reply:
The base interface only covers for Collection, so I think it is debatable. Is there a way to provide an extra compatibility "overloaded" method? I think I saw this somewhere.

Reply:
Normally the submitter would send a pull request to the plugin to make it compatible, as we can't merge this if it breaks the compatibility tests across the eco system.

Is that something you could do? 

In the linked case its probably a matter of changing it from List to Collection

Reply:
Isn't this a use case for the adapter in https://github.com/jenkinsci/bridge-method-injector?

Reply:
Isn't this just plain binary incompatibility, so the JUnit Attachments plugin would just need a rebuild? 

It seems it is a case for the Bridge Method Injector.

Reply:
it can be temporarily but afaik they are a bit of a nightmare, IDEs don't work well with them at all.

@basil has tried it somewhat recently and I think decided against it.

Reply:
The plugin is casting to a list, but its actually a set and they don't have a common interface so no I don't think a rebuild would help at all

Reply:
https://github.com/jenkinsci/junit-attachments-plugin/pull/127

If we decide to add a bridge method, I have prepared some code to try it, but overall I don't really know how to test it properly and my head is already spinning.

Reply:
I really do not recommend using bridge methods if you can possibly avoid it. Are there any plugins calling this override directly from production (not test) code? If so, the safest approach would be to deprecate the interface method and introduce `getChildren2` or some nicer name.

Reply:
This is the only place I could find: https://github.com/jenkinsci/testng-plugin-plugin/blob/656853e1ee41bb7890d74924c5537113be3b0939/src/main/java/hudson/plugins/testng/results/TestNGResult.java#L194

Reply:
That is `PackageResult.getChildren` not `ClassResult.getChildren`?

Reply:
If there _are_ any callers of the override from production code, and they do not need the children subtyped to `CaseResult`, then you could prepare the caller proactively by first assigning the `ClassResult` to the superclass to make sure the bytecode refers to the more generic method. Such a patch would need to be released in advanced, and users advised to apply the update before updating `junit`.

Reply:
Right and thats already a collection which the plugin is assuming is a list, but thats a separate issue.

Reply:
Still failing:
https://ci.jenkins.io/blue/organizations/jenkins/Tools%2Fbom/detail/PR-3394/3/tests

```
java.lang.NoSuchMethodError: 'java.util.List hudson.tasks.junit.ClassResult.getChildren()'
	at hudson.plugins.junitattachments.AttachmentPublisherPipelineTest.annotationDoesNotFailForPipeline(AttachmentPublisherPipelineTest.java:82)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:655)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

bit confused though as that was changed to a collection:
```
Collection<CaseResult> caseResults = cr.getChildren();
```

Reply:
Not sure what needs to be done now.

Reply:
> bit confused though as that was changed to a collection

Reply:
> bit confused though as that was changed to a collection

Well the variable it was assigned to was changed to a collection. The bytecode is still calling `List ClassResult.getChildren()`, because it was compiled against the current `junit` release. This can only be solved by calling `getChildren` on a receiver of static type given by the supertype declared to return `Collection`.

Reply:
(One of the times where this behavior of PCT is genuinely helpful: it forcibly reminds you of the difference between source and binary compatibility.)

Reply:
so introduce `getChildren2` with a deprecation of the old?

Reply:
If necessary. If there are no known callers of the current override from production code, I think it would be OK to break binary compatibility so long as tests are prepared like in https://github.com/jenkinsci/junit-plugin/pull/625#discussion_r1688184691.

Reply:
There are no known consumers that I could find, so if I filed a PR for an incremental against junit-attachments and then bom it should work?

Reply:
BTW is it really necessary to change this signature?

> this is a performance critical area

Understood, but where specifically is the performance-critical call site? If it is within this plugin, you could simply add a new method (`@Restricted(NoExternalUse.class)`) that returns a more helpful type and call that internally.

Reply:
That is a risky approach since that PR cannot be released until this PR has been.

Reply:
Try https://github.com/jenkinsci/junit-attachments-plugin/pull/129 in the `bom`.

Reply:
> BTW is it really necessary to change this signature?
> 
> > this is a performance critical area
> 
> Understood, but where specifically is the performance-critical call site? If it is within this plugin, you could simply add a new method (`@Restricted(NoExternalUse.class)`) that returns a more helpful type and call that internally.

I think we should break this. The base interface only provides a Collection. 

Now, regarding the performance: TL;DR is that there is recursion when navigating the results and it really shows when doing so on 10K test results with multiple subsets. 

More on the reasons

- see getSafeName, uniquifyName; every time navigating the results it is used to determine if the name is unique enough 
- on why it's now a TreeSet instead of a List: it helps when loading the junitResult.xml by loading it by sorting on insert instead of after the fact, which is simpler

We are looking at performance regressions on top of other performance regressions. I can barely read the code, I think we should avoid adding even more compatibility functions when all we have is basically "monolithic" plugin code (having a non-interface used by some external extension code is monolithic to me).

Reply:
I can also apply the bridge method injector to see if it does anything. Better to have something in place. It's basically a "two liner"... 

I am curious, I don't know how to test it locally; not that into the whereabouts of plugin repositories in Jenkins yet. 

We can always revert it.

Reply:
lets avoid it if we can, its a pain to develop with

Reply:
Ok, PCT passed with @jglick's fix which has been released in https://github.com/jenkinsci/junit-attachments-plugin/pull/129